contracts/BookMarketplace.sol

```diff
@@ -290,10 +309,76 @@ contract BookMarketplace is ERC721URIStorage {
     }
 
     /* get the listing for a given token id */
-    function getListingByTokenId(uint256 tokenId) public view returns (Listing memory) {
+    function getListingByTokenId(
+        uint256 tokenId
+    ) public view returns (Listing memory) {
         return idToMarketItem[tokenId];
     }
 
+    /* Creates the sale of a marketplace item */
+    /* Transfers ownership of the item, as well as funds between parties */
+    function createMarketSale(uint256 tokenId) public payable {
+        uint price = idToMarketItem[tokenId].instantPrice;
+        address seller = idToMarketItem[tokenId].seller;
+        require(
+            msg.value == price,
+            "Please submit the asking price in order to complete the purchase"
+        );
+        idToMarketItem[tokenId].owner = payable(msg.sender);
+        idToMarketItem[tokenId].sold = true;
+        idToMarketItem[tokenId].seller = payable(address(0));
+        _itemsSold.increment();
+        _transfer(address(this), msg.sender, tokenId);
+        payable(marketOwner).transfer(listingPrice);
+        payable(seller).transfer(msg.value);
+    }
+
+    /* Creates the sale of a marketplace item */
+    /* Transfers ownership of the item, as well as funds between parties */
+    function createBiddingSale(
+        uint256 tokenId
+    ) public onlyListingSeller(tokenId) {
+        require(
+            idToMarketItem[tokenId].allowBid,
+            "This listing does not allow bidding"
+        );
+
+        int256 highestBidIndex = getHighestBidIndex(tokenId);
+        require(highestBidIndex > 0, "This listing does not have any bid yet");
+
+        //Mark as chosen
+        idToMarketItem[tokenId].bidList[uint256(highestBidIndex)].chosen = true;
+    }
+
+    function acceptBiddingSale(
+        uint256 tokenId,
+        uint256 bidIndex
+    ) public payable {
+        Bid memory bid = idToMarketItem[tokenId].bidList[bidIndex];
+        require(bid.chosen, "Bid has not been chosen");
+        require(bid.bidder == msg.sender, "You are not the bidder");
+        require(
+            bid.bidAmount == msg.value,
+            "Please submit the price equals the bid amount"
+        );
+
+        //Transfer bid price
+        address payable seller = idToMarketItem[tokenId].seller;
+        (bool success, ) = seller.call{value: msg.value}("");
+        require(success, "Transfer failed.");
```

Comment:
As an aside: we should probably figure out some kind of callback function that is appropriate for using .call, at some point. (no follow-up needed)